### DISCUSSION TITLE
[SIP] Add More Deck.gl Map Layers

### DISCUSSION THREAD

**Discussion by Mattc1221** [apache/superset#23553] [General]
### Motivation

Superset already has supported functionality for other deck.gl visualization layers (e.g. Arc, Geojson, Grid, Hex, Path, Polygon, Scatter, Screengrid). Since Superset already contains the logic to handle deck.gl layers, it is straightforward and easy to maintain to add new deck.gl layers.

By expanding the range of available deck.gl visualization layers, users will have more options to choose from when creating their visualizations. This will allow them to tailor their visualizations to their specific needs and explore their data in different ways. It will also allow Superset to better compete with other data visualization tools on the market that offer a wider range of visualization options.

Overall, I think adding more deck.gl layers to Superset is a great idea and has the potential to significantly improve the user experience and the insights that users can gain from their data.

Noted in https://github.com/apache/superset/discussions/22930 as well

### Proposed Change

Adding the following deck.gl mapping layers as chart types:

- [HeatmapLayer](https://deck.gl/docs/api-reference/aggregation-layers/heatmap-layer)
- [ContourLayer](https://deck.gl/docs/api-reference/aggregation-layers/contour-layer)
- [TripsLayer](https://deck.gl/docs/api-reference/geo-layers/trips-layer)

<details>

<summary>Example Interface for new deck.gl heatmap interface</summary>

![Screenshot 2023-04-01 at 11 58 28 PM](https://user-images.githubusercontent.com/36670322/229332263-f7a45b38-ffa4-455b-84a3-492cadfcb8db.png)
![Screenshot 2023-04-02 at 12 00 19 AM](https://user-images.githubusercontent.com/36670322/229332332-f5e6603b-b92c-4669-89ef-5606c1e09485.png)

</details>

### New or Changed Public Interfaces

The new chart types will be added under the Map category for charts. Each chart type will have a new front-end component and back-end class inherited from BaseDeckGLViz.

### New dependencies

This change will not require any new dependencies since the deck.gl library is already installed in the system.

### Migration Plan and Compatibility

No migrations will be necessary since the added deck.gl layers will be treated as new chart types.

### Rejected Alternatives

Alternative approaches considered and rejected include using other mapping libraries and not adding new mapping layers. We rejected the option of using other mapping libraries because using deck.gl visualizations within our codebase helps create a consistent interface for the user and reduces the technical overhead in the debugging and feature developing process. We rejected the option of not adding new mapping layers because adding more options for users to display their data allows for better data stories, thus making superset a better tool overall.

### Pull Requests:

- [x] deck.gl Heatmap Layer PR: https://github.com/apache/superset/pull/23551
- [ ] deck.gl Contour Layer PR: https://github.com/apache/superset/pull/24154
- [ ] deck.gl Trips Layer PR: N/A

**Comment by DavidGonzalez-Earth** [comment]
Hi! I would find this very useful, especially the TripsLayer, as this is functionality we were seeking to provide to our end users in the near term. We deal with road transportation and logistics, currently using the arc layer to display trips, but would like to the actual route on a map in superset, as this provides much more detail.

**Comment by villebro** [comment]
It would be great to migrate the DeckGL plugin to the v1 chart data API and be able to drop the "legacy" prefix from the plugin name. I can take a stab at seeing how much work that would require.

**Comment by rusackas** [comment]
I love this, and applaud the improvements @Mattc1221 is making with these new layers! It's _awesome_

That said, I feel a need to wave a tiny flag that these do add to the maintenance footprint of an already problematic set of plugins. To @villebro 's point, we need to modernize these plugins. Upgrading their architecture (namely that endpoint) will better cement their future in the codebase, rather than force their eventual deprecation and removal. There are also _numerous_ bugs and other issues that need to be addressed in the existing plugins. If anyone seeing this thread is able to undertake some of these tasks, it would be a _hugely_ appreciated effort.

**Comment by Mattc1221** [comment]
Hey @rusackas and @villebro! I really like the deck.gl plugin and would love to start helping on migrating the plugin to use the v1 chart data API. I'm wondering if you know of any good starting points for this migration process (i.e. what kind of steps need to be taken, and some documentation on the chart data api). Otherwise, if you think it would be simpler to meet, please message me over slack and we can set up a virtual meeting!

**Reply by Yisel-Hernandez** [reply]
Lo estoy implementando pero lo que pasa es que estoy tratando de filtrar un gráfico de multiples capas, la idea que tuve fue en el mismo dashboard agregar los gráficos que componen el gráfico de múltiples capas, pero solo se filtran los gráficos por aparte pero no afecta en nada el de múltiples capas.

Alguno tiene idea de cómo podría hacer para que a medida de que se filtren los gráficos que componen el gráfico de múltiples capas este también se me filtre con los cambios de los demás gráficos?

Muchas gracias.
